From 79e878aefe85f0c921c954821903c8cea3652bc9 Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Thu, 25 Jun 2026 14:46:23 +0300
Subject: [PATCH] try fixing openvino theading detection

---
 ggml/src/ggml-openvino/CMakeLists.txt | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/ggml/src/ggml-openvino/CMakeLists.txt b/ggml/src/ggml-openvino/CMakeLists.txt
index cc089b72..ee41dd85 100644
--- a/ggml/src/ggml-openvino/CMakeLists.txt
+++ b/ggml/src/ggml-openvino/CMakeLists.txt
@@ -1,5 +1,13 @@
 find_package(OpenVINO REQUIRED COMPONENTS Runtime Threading)
+find_package(OpenVINO REQUIRED COMPONENTS Runtime)
 find_package(OpenCL REQUIRED)
+find_package(TBB REQUIRED)
+
+if(NOT TARGET openvino::threading)
+    add_library(openvino_threading INTERFACE)
+    target_link_libraries(openvino_threading INTERFACE TBB::tbb)
+    add_library(openvino::threading ALIAS openvino_threading)
+endif()
 
 file(GLOB_RECURSE GGML_HEADERS_OPENVINO "*.h" "*.hpp")
 file(GLOB_RECURSE GGML_SOURCES_OPENVINO "*.cpp")
-- 
2.52.0.windows.1

