From 1ef75561a874dfef7b3bdb188c2580bb0351cb00 Mon Sep 17 00:00:00 2001
From: rhythmcache <153998419+rhythmcache@users.noreply.github.com>
Date: Thu, 28 Aug 2025 07:46:24 -0700
Subject: [PATCH] renamed bpf to codec2_bpf

---
 src/bpf.h   | 2 +-
 src/c2sim.c | 4 ++--
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/bpf.h b/src/bpf.h
index d74e059b..e03e3e2e 100644
--- a/src/bpf.h
+++ b/src/bpf.h
@@ -1,6 +1,6 @@
 #define BPF_N 101
 
-float bpf[] = {
+float codec2_bpf[] = {
     0.002174,  0.003245,  0.002147,  0.001866,  0.002764,  0.000567,  -0.001641,
     -0.000565, -0.002415, -0.005837, -0.003620, -0.002828, -0.006268, -0.002787,
     0.001963,  -0.001234, 0.001446,  0.009200,  0.005331,  0.003521,  0.011821,
diff --git a/src/c2sim.c b/src/c2sim.c
index 31174157..fd9d3c8e 100644
--- a/src/c2sim.c
+++ b/src/c2sim.c
@@ -149,7 +149,7 @@ int main(int argc, char *argv[]) {
       {"vq_pitch_e", no_argument, &vector_quant_Wo_e, 1},
       {"rate", required_argument, NULL, 0},
       {"gain", required_argument, NULL, 0},
-      {"bpf", no_argument, &bpf_en, 1},
+      {"codec2_bpf", no_argument, &bpf_en, 1},
       {"bpfb", no_argument, &bpfb_en, 1},
       {"amread", required_argument, &amread, 1},
       {"hmread", required_argument, &hmread, 1},
@@ -628,7 +628,7 @@ int main(int argc, char *argv[]) {
       if (bpfb_en)
         inverse_filter(&bpf_buf[BPF_N], bpfb, N_SAMP, buf_float, BPF_N);
       else
-        inverse_filter(&bpf_buf[BPF_N], bpf, N_SAMP, buf_float, BPF_N);
+        inverse_filter(&bpf_buf[BPF_N], codec2_bpf, N_SAMP, buf_float, BPF_N);
     }
 
     /* shift buffer of input samples, and insert new samples */
