From 6e0a0e09c065aa5401eb9c30d724240fffe890f1 Mon Sep 17 00:00:00 2001
From: rhythmcache <153998419+rhythmcache@users.noreply.github.com>
Date: Fri, 3 Oct 2025 19:13:54 -0700
Subject: [PATCH] Update GetDependencies.cmake.in

---
 cmake/GetDependencies.cmake.in | 12 +-----------
 1 file changed, 1 insertion(+), 11 deletions(-)

diff --git a/cmake/GetDependencies.cmake.in b/cmake/GetDependencies.cmake.in
index 0d25f670..fad02c42 100644
--- a/cmake/GetDependencies.cmake.in
+++ b/cmake/GetDependencies.cmake.in
@@ -11,14 +11,4 @@ set(CMAKE_SYSTEM_LIBRARY_PATH @CMAKE_SYSTEM_LIBRARY_PATH@)
 set(CMAKE_FIND_ROOT_PATH @CMAKE_FIND_ROOT_PATH@)
 set(CODEC2_DLL ${CMAKE_BINARY_DIR}/src/libcodec2.dll)
 
-include(${CMAKE_SOURCE_DIR}/cmake/GetPrerequisites.cmake)
-get_prerequisites(${CODEC2_DLL} _deps 1 0 "" "")
-foreach(_runtime ${_deps})
-    message("Looking for ${_runtime}")
-    find_library(RUNTIME_${_runtime} ${_runtime})
-    message("${RUNTIME_${_runtime}}")
-    if(RUNTIME_${_runtime})
-        file(INSTALL DESTINATION "${CMAKE_INSTALL_PREFIX}/bin"
-        TYPE EXECUTABLE FILES "${RUNTIME_${_runtime}}")
-    endif()
-endforeach()
+
