From 4de9dafe3350eafc975c63d8e5ce438f8038caa7 Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Sat, 11 Jul 2026 15:17:25 +0300
Subject: [PATCH] disable tests

---
 Makefile | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/Makefile b/Makefile
index f347162..28aed5d 100644
--- a/Makefile
+++ b/Makefile
@@ -149,8 +149,6 @@ clean:
 
 # Validate
 
-check: all
-	$(MAKE) -C tests
 
 reflow:
 	@clang-format --style="{IndentWidth: 8, UseTab: ForIndentation}" -i $$(find . -name "*.[ch]")
@@ -229,7 +227,7 @@ EXTRAS =     NEWS \
 	     control \
 	     local.dic
 
-ALL =  Makefile *.[ch] *.adoc doc tests pic $(EXTRAS)
+ALL =  Makefile *.[ch] *.adoc doc pic $(EXTRAS)
 
 giflib-$(VERSION).tar.gz: $(ALL)
 	mkdir giflib-$(VERSION)
-- 
2.52.0.windows.1

