From 4a80ec98641ddc08b0a6465967823df4d63de858 Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Tue, 2 Jun 2026 07:36:30 +0300
Subject: [PATCH] fix for improved the pthread shim used on Windows

---
 libvmaf/src/feature/integer_vif.c | 28 ++++++++++++++--------------
 1 file changed, 14 insertions(+), 14 deletions(-)

diff --git a/src/feature/integer_vif.c b/src/feature/integer_vif.c
index 069423ad..1077ae03 100644
--- a/src/feature/integer_vif.c
+++ b/src/feature/integer_vif.c
@@ -644,20 +644,20 @@ static int init(VmafFeatureExtractor *fex, enum VmafPixelFormat pix_fmt,
     s->public.buf.data = data; data += pad_size;
     s->public.buf.ref = data; data += frame_size + pad_size + pad_size;
     s->public.buf.dis = data; data += frame_size + pad_size;
-    s->public.buf.mu1 = data; data += h * s->public.buf.stride_16;
-    s->public.buf.mu2 = data; data += h * s->public.buf.stride_16;
-    s->public.buf.mu1_32 = data; data += s->public.buf.stride_32;
-    s->public.buf.mu2_32 = data; data += s->public.buf.stride_32;
-    s->public.buf.ref_sq = data; data += s->public.buf.stride_32;
-    s->public.buf.dis_sq = data; data += s->public.buf.stride_32;
-    s->public.buf.ref_dis = data; data += s->public.buf.stride_32;
-    s->public.buf.tmp.mu1 = data; data += s->public.buf.stride_tmp;
-    s->public.buf.tmp.mu2 = data; data += s->public.buf.stride_tmp;
-    s->public.buf.tmp.ref = data; data += s->public.buf.stride_tmp;
-    s->public.buf.tmp.dis = data; data += s->public.buf.stride_tmp;
-    s->public.buf.tmp.ref_dis = data; data += s->public.buf.stride_tmp;
-    s->public.buf.tmp.ref_convol = data; data += s->public.buf.stride_tmp;
-    s->public.buf.tmp.dis_convol = data;
+    s->public.buf.mu1 = (uint16_t *)data; data += h * s->public.buf.stride_16;
+    s->public.buf.mu2 = (uint16_t *)data; data += h * s->public.buf.stride_16;
+    s->public.buf.mu1_32 = (uint32_t *)data; data += s->public.buf.stride_32;
+    s->public.buf.mu2_32 = (uint32_t *)data; data += s->public.buf.stride_32;
+    s->public.buf.ref_sq = (uint32_t *)data; data += s->public.buf.stride_32;
+    s->public.buf.dis_sq = (uint32_t *)data; data += s->public.buf.stride_32;
+    s->public.buf.ref_dis = (uint32_t *)data; data += s->public.buf.stride_32;
+    s->public.buf.tmp.mu1 = (uint32_t *)data; data += s->public.buf.stride_tmp;
+    s->public.buf.tmp.mu2 = (uint32_t *)data; data += s->public.buf.stride_tmp;
+    s->public.buf.tmp.ref = (uint32_t *)data; data += s->public.buf.stride_tmp;
+    s->public.buf.tmp.dis = (uint32_t *)data; data += s->public.buf.stride_tmp;
+    s->public.buf.tmp.ref_dis = (uint32_t *)data; data += s->public.buf.stride_tmp;
+    s->public.buf.tmp.ref_convol = (uint32_t *)data; data += s->public.buf.stride_tmp;
+    s->public.buf.tmp.dis_convol = (uint32_t *)data;
 
     s->feature_name_dict =
         vmaf_feature_name_dict_from_provided_features(fex->provided_features,
-- 
2.52.0.windows.1

