From 84cb0895627e0ee371d4dfad60f3c81dc0586527 Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Mon, 13 Jul 2026 05:45:26 +0300
Subject: [PATCH] disable tests

---
 CMakeLists.txt | 188 +------------------------------------------------
 1 file changed, 3 insertions(+), 185 deletions(-)

diff --git a/third_party/highway/CMakeLists.txt b/third_party/highway/CMakeLists.txt
index ea8b330c..4dd41470 100644
--- a/third_party/highway/CMakeLists.txt
+++ b/third_party/highway/CMakeLists.txt
@@ -208,13 +208,6 @@ if (NOT HWY_CMAKE_HEADER_ONLY)
   )
 endif()
 
-set(HWY_TEST_SOURCES
-    hwy/tests/hwy_gtest.h
-    hwy/tests/test_util-inl.h
-    hwy/tests/test_util.cc
-    hwy/tests/test_util.h
-)
-
 if (MSVC)
   set(HWY_FLAGS
     # fix build error C1128 in blockwise*_test & arithmetic_test
@@ -480,7 +473,7 @@ if(UNIX AND NOT APPLE)
 endif()
 endif()  # HWY_ENABLE_CONTRIB
 
-add_library(hwy_test ${HWY_LIBRARY_TYPE} ${HWY_TEST_SOURCES})
+add_library(hwy_test ${HWY_LIBRARY_TYPE})
 target_link_libraries(hwy_test PUBLIC hwy)
 target_compile_options(hwy_test PRIVATE ${HWY_FLAGS})
 set_property(TARGET hwy_test PROPERTY POSITION_INDEPENDENT_CODE ON)
@@ -503,7 +496,6 @@ endif()
 # -------------------------------------------------------- hwy_list_targets
 # Generate a tool to print the compiled-in targets as defined by the current
 # flags. This tool will print to stderr at build time, after building hwy.
-add_executable(hwy_list_targets hwy/tests/list_targets.cc)
 target_compile_options(hwy_list_targets PRIVATE ${HWY_FLAGS})
 target_compile_features(hwy_list_targets PRIVATE ${HWY_CXX_STD_TGT_COMPILE_FEATURE})
 target_link_libraries(hwy_list_targets PRIVATE hwy)
@@ -555,31 +547,13 @@ foreach (source ${HWY_CONTRIB_SOURCES})
 endforeach()
 endif()  # HWY_ENABLE_CONTRIB
 
-if (HWY_ENABLE_TESTS)
-install(TARGETS hwy_test EXPORT hwy_targets
-  LIBRARY DESTINATION "${CMAKE_INSTALL_LIBDIR}"
-  ARCHIVE DESTINATION "${CMAKE_INSTALL_LIBDIR}"
-  RUNTIME DESTINATION "${CMAKE_INSTALL_BINDIR}")
-# Install all the headers keeping the relative path to the current directory
-# when installing them.
-foreach (source ${HWY_TEST_SOURCES})
-  if ("${source}" MATCHES "\.h$")
-    get_filename_component(dirname "${source}" DIRECTORY)
-    install(FILES "${source}"
-        DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}/${dirname}")
-  endif()
-endforeach()
-endif()  # HWY_ENABLE_TESTS
-
 # Add a pkg-config file for libhwy and the contrib/test libraries.
 set(HWY_LIBRARY_VERSION "${CMAKE_PROJECT_VERSION}")
 set(HWY_PC_FILES libhwy.pc)
 if (HWY_ENABLE_CONTRIB)
 list(APPEND HWY_PC_FILES libhwy-contrib.pc)
 endif()  # HWY_ENABLE_CONTRIB
-if (HWY_ENABLE_TESTS)
-list(APPEND HWY_PC_FILES libhwy-test.pc)
-endif()  # HWY_ENABLE_TESTS
+
 foreach (pc ${HWY_PC_FILES})
   configure_file("${CMAKE_CURRENT_SOURCE_DIR}/${pc}.in" "${pc}" @ONLY)
   install(FILES "${CMAKE_CURRENT_BINARY_DIR}/${pc}"
@@ -588,73 +562,8 @@ endforeach()
 
 endif()  # HWY_ENABLE_INSTALL
 # -------------------------------------------------------- Examples
-if (HWY_ENABLE_EXAMPLES)
-
-# Avoids mismatch between GTest's static CRT and our dynamic.
-set(gtest_force_shared_crt ON CACHE BOOL "" FORCE)
-
-# Programming exercise with integrated benchmark
-add_executable(hwy_benchmark hwy/examples/benchmark.cc)
-target_sources(hwy_benchmark PRIVATE
-    hwy/nanobenchmark.h)
-# Try adding one of -DHWY_COMPILE_ONLY_SCALAR, -DHWY_COMPILE_ONLY_EMU128 or
-# -DHWY_COMPILE_ONLY_STATIC to observe the difference in targets printed.
-target_compile_options(hwy_benchmark PRIVATE ${HWY_FLAGS})
-target_compile_features(hwy_benchmark PRIVATE ${HWY_CXX_STD_TGT_COMPILE_FEATURE})
-target_link_libraries(hwy_benchmark PRIVATE hwy)
-target_link_libraries(hwy_benchmark PRIVATE ${ATOMICS_LIBRARIES})
-set_target_properties(hwy_benchmark
-    PROPERTIES RUNTIME_OUTPUT_DIRECTORY "examples/")
-
-# Profiler demo
-add_executable(hwy_profiler_example hwy/examples/profiler_example.cc)
-target_sources(hwy_profiler_example PRIVATE
-    hwy/profiler.h)
-target_compile_options(hwy_profiler_example PRIVATE ${HWY_FLAGS})
-target_compile_features(hwy_profiler_example PRIVATE ${HWY_CXX_STD_TGT_COMPILE_FEATURE})
-target_link_libraries(hwy_profiler_example PRIVATE hwy)
-target_link_libraries(hwy_profiler_example PRIVATE ${ATOMICS_LIBRARIES})
-set_target_properties(hwy_profiler_example
-    PROPERTIES RUNTIME_OUTPUT_DIRECTORY "examples/")
-
-endif()  # HWY_ENABLE_EXAMPLES
-# -------------------------------------------------------- Tests
-
-include(CTest)
-
-if(BUILD_TESTING AND HWY_ENABLE_TESTS)
-enable_testing()
-include(GoogleTest)
 
-set(HWY_SYSTEM_GTEST OFF CACHE BOOL "Use pre-installed googletest?")
-if(HWY_SYSTEM_GTEST)
-find_package(GTest REQUIRED)
-else()
-# Download and unpack googletest at configure time
-configure_file(CMakeLists.txt.in googletest-download/CMakeLists.txt)
-execute_process(COMMAND ${CMAKE_COMMAND} -G "${CMAKE_GENERATOR}" .
-  RESULT_VARIABLE result
-  WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/googletest-download )
-if(result)
-  message(FATAL_ERROR "CMake step for googletest failed: ${result}")
-endif()
-execute_process(COMMAND ${CMAKE_COMMAND} --build .
-  RESULT_VARIABLE result
-  WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/googletest-download )
-if(result)
-  message(FATAL_ERROR "Build step for googletest failed: ${result}")
-endif()
-
-# Prevent overriding the parent project's compiler/linker
-# settings on Windows
-set(gtest_force_shared_crt ON CACHE BOOL "" FORCE)
-
-# Add googletest directly to our build. This defines
-# the gtest and gtest_main targets.
-add_subdirectory(${CMAKE_CURRENT_BINARY_DIR}/googletest-src
-                 ${CMAKE_CURRENT_BINARY_DIR}/googletest-build
-                 EXCLUDE_FROM_ALL)
-endif()  # HWY_SYSTEM_GTEST
+# -------------------------------------------------------- Tests
 
 set(HWY_TEST_FILES
   hwy/contrib/algo/copy_test.cc
@@ -667,53 +576,6 @@ set(HWY_TEST_FILES
   hwy/highway_test.cc
   hwy/nanobenchmark_test.cc
   hwy/targets_test.cc
-  hwy/examples/skeleton_test.cc
-  hwy/tests/arithmetic_test.cc
-  hwy/tests/bit_permute_test.cc
-  hwy/tests/blockwise_shift_test.cc
-  hwy/tests/blockwise_test.cc
-  hwy/tests/cast_test.cc
-  hwy/tests/combine_test.cc
-  hwy/tests/compare_test.cc
-  hwy/tests/compress_test.cc
-  hwy/tests/convert_test.cc
-  hwy/tests/count_test.cc
-  hwy/tests/crypto_test.cc
-  hwy/tests/demote_test.cc
-  hwy/tests/div_test.cc
-  hwy/tests/dup128_vec_test.cc
-  hwy/tests/expand_test.cc
-  hwy/tests/float_test.cc
-  hwy/tests/foreach_vec_test.cc
-  hwy/tests/if_test.cc
-  hwy/tests/in_range_float_to_int_conv_test.cc
-  hwy/tests/interleaved_test.cc
-  hwy/tests/logical_test.cc
-  hwy/tests/mask_combine_test.cc
-  hwy/tests/mask_convert_test.cc
-  hwy/tests/mask_mem_test.cc
-  hwy/tests/mask_slide_test.cc
-  hwy/tests/mask_test.cc
-  hwy/tests/masked_arithmetic_test.cc
-  hwy/tests/masked_minmax_test.cc
-  hwy/tests/memory_test.cc
-  hwy/tests/minmax_test.cc
-  hwy/tests/mul_test.cc
-  hwy/tests/reduction_test.cc
-  hwy/tests/resize_test.cc
-  hwy/tests/reverse_test.cc
-  hwy/tests/saturated_test.cc
-  hwy/tests/shift_test.cc
-  hwy/tests/shuffle4_test.cc
-  hwy/tests/slide_up_down_test.cc
-  hwy/tests/sums_abs_diff_test.cc
-  hwy/tests/swizzle_block_test.cc
-  hwy/tests/swizzle_test.cc
-  hwy/tests/table_test.cc
-  hwy/tests/test_util_test.cc
-  hwy/tests/truncate_test.cc
-  hwy/tests/tuple_test.cc
-  hwy/tests/widen_mul_test.cc
 )
 
 set(HWY_TEST_LIBS hwy hwy_test)
@@ -738,50 +600,6 @@ list(APPEND HWY_TEST_FILES
 )
 endif()  # HWY_ENABLE_CONTRIB
 
-if(HWY_SYSTEM_GTEST)
-  if (CMAKE_VERSION VERSION_LESS 3.20)
-    set(HWY_GTEST_LIBS GTest::GTest GTest::Main)
-  else()
-    set(HWY_GTEST_LIBS GTest::gtest GTest::gtest_main)
-  endif()
-else()
-  set(HWY_GTEST_LIBS gtest gtest_main)
-endif()
-
-file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/tests)
-foreach (TESTFILE IN LISTS HWY_TEST_FILES)
-  # The TESTNAME is the name without the extension or directory.
-  get_filename_component(TESTNAME ${TESTFILE} NAME_WE)
-  add_executable(${TESTNAME} ${TESTFILE})
-  target_compile_options(${TESTNAME} PRIVATE ${HWY_FLAGS})
-  # Test all targets, not just the best/baseline. This changes the default
-  # policy to all-attainable; note that setting -DHWY_COMPILE_* directly can
-  # cause compile errors because only one may be set, and other CMakeLists.txt
-  # that include us may set them.
-  target_compile_options(${TESTNAME} PRIVATE -DHWY_IS_TEST=1)
-  target_compile_features(${TESTNAME} PRIVATE ${HWY_CXX_STD_TGT_COMPILE_FEATURE})
-
-  target_link_libraries(${TESTNAME} PRIVATE ${HWY_TEST_LIBS} ${HWY_GTEST_LIBS})
-  # For GCC __atomic_store_8, see #887
-  target_link_libraries(${TESTNAME} PRIVATE ${ATOMICS_LIBRARIES})
-  # Output test targets in the test directory.
-  set_target_properties(${TESTNAME} PROPERTIES RUNTIME_OUTPUT_DIRECTORY "tests")
-
-  if (HWY_EMSCRIPTEN)
-    set_target_properties(${TESTNAME} PROPERTIES LINK_FLAGS "-s SINGLE_FILE=1")
-  endif()
-
-  if(${CMAKE_VERSION} VERSION_LESS "3.10.3")
-    gtest_discover_tests(${TESTNAME} TIMEOUT 60)
-  else ()
-    gtest_discover_tests(${TESTNAME} DISCOVERY_TIMEOUT 60)
-  endif ()
-endforeach ()
-
-# The skeleton test uses the skeleton library code.
-target_sources(skeleton_test PRIVATE hwy/examples/skeleton.cc)
-
-endif()  # BUILD_TESTING
 
 if (HWY_ENABLE_INSTALL)
   # write hwy-config file to handle `Config` mode
-- 
2.52.0.windows.1

