From 8b7471eab8d09536b3c104dbb30a65699cf48104 Mon Sep 17 00:00:00 2001
From: vfjpl <vfjpl1@gmail.com>
Date: Wed, 16 Jul 2025 16:05:18 +0200
Subject: [PATCH] Fix ioctlsocket
---
--- a/librabbitmq/amqp_socket.c
+++ b/librabbitmq/amqp_socket.c
@@ -282,7 +282,10 @@
 
 #ifdef _WIN32
 static int connect_socket(struct addrinfo *addr, amqp_time_t deadline) {
-  int one = 1;
+  u_long one = 1;
+  /* setsockopt expects (const char*) on Windows, but let's keep a separate 
+     int for those calls to avoid any potential issues with sizeof */
+  int opt_one = 1; 
   SOCKET sockfd;
   int last_error;
 
@@ -291,14 +294,14 @@
    * http://stackoverflow.com/questions/1953639/is-it-safe-to-cast-socket-to-int-under-win64
    */
 
-  sockfd = (int)socket(addr->ai_family, addr->ai_socktype, addr->ai_protocol);
+  sockfd = socket(addr->ai_family, addr->ai_socktype, addr->ai_protocol);
   if (INVALID_SOCKET == sockfd) {
     return AMQP_STATUS_SOCKET_ERROR;
   }
 
   /* Set the socket to be non-blocking */
   if (SOCKET_ERROR == ioctlsocket(sockfd, FIONBIO, &one)) {
-    last_error = AMQP_STATUS_SOCKET_ERROR;
+    last_error = AMQP_STATUS_SOCKET_ERROR;
     goto err;
   }
 
@@ -306,13 +309,13 @@
   if (SOCKET_ERROR == setsockopt(sockfd, IPPROTO_TCP, TCP_NODELAY,
-                                 (const char *)&one, sizeof(one))) {
+                                 (const char *)&opt_one, sizeof(opt_one))) {
     last_error = AMQP_STATUS_SOCKET_ERROR;
     goto err;
   }
 
   /* Enable TCP keepalives */
   if (SOCKET_ERROR == setsockopt(sockfd, SOL_SOCKET, SO_KEEPALIVE,
-                                 (const char *)&one, sizeof(one))) {
+                                 (const char *)&opt_one, sizeof(opt_one))) {
     last_error = AMQP_STATUS_SOCKET_ERROR;
     goto err;
   }
 
-- 
2.52.0.windows.1