From 288e56e959fa2123cc3da8f0b0ba61f62e57b105 Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Tue, 21 Apr 2026 03:59:48 +0300
Subject: [PATCH] use external mbedtls

---
 meson.build | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/meson.build b/meson.build
index bfee12b..8c2a602 100755
--- a/meson.build
+++ b/meson.build
@@ -250,8 +250,6 @@ crypto_deps = []
 
 mbedcrypto_lib_found = false
 if use_mbedtls
-	message('Building mbedtls')
-	subdir('contrib/mbedtls')
-
-	crypto_deps += mbedcrypto_lib
+	mbedcrypto_lib_found = true
+	message('Using external mbedtls')
 elif use_gnutls
-- 
2.52.0.windows.1