From 28deae4d211f6f8d923acd5d4357a91c33c0a6e9 Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Thu, 18 Jun 2026 13:01:44 +0300
Subject: [PATCH] disable nettle search

---
 meson.build | 32 +++++++++++---------------------
 1 file changed, 11 insertions(+), 21 deletions(-)

diff --git a/meson.build b/meson.build
index 20fa42a..92b57d9 100755
--- a/meson.build
+++ b/meson.build
@@ -291,17 +291,7 @@ if use_mbedtls
 
 	crypto_deps += mbedcrypto_lib
 elif use_gnutls
-	crypto_deps = [dependency('nettle'), dependency('hogweed')]
-	test_mini_gmp = '''
-#include <nettle/bignum.h>
-#if !NETTLE_USE_MINI_GMP
-	#error "Need to link gmp"
-#endif
-'''
-	if not cc.compiles(test_mini_gmp, name: 'Nettle with mini-gmp?', dependencies: crypto_deps)
-		crypto_deps += dependency('gmp')
-	endif
-	crypto_deps += dependency('gnutls')#Just for random numbers at the moment :/
+    crypto_deps = [dependency('gnutls')]
 endif
 deps += crypto_deps
 
@@ -323,26 +313,26 @@ if get_option('allow_obj_filter') and get_option('default_library') != 'shared'
 	endif
 endif
 
-if not mbedcrypto_lib_found and not use_gnutls
-	platform_files += [
-		'contrib/aes.c',
-		'contrib/sha256.c',
-		'contrib/fastpbkdf2.c',
-	]
+if not mbedcrypto_lib_found
+    platform_files += [
+        'contrib/aes.c',
+        'contrib/sha256.c',
+        'contrib/fastpbkdf2.c',
+    ]
 endif
 
 have_srp = mbedcrypto_lib_found or use_gnutls
-
 platform_files += 'src/crypto/random.c'
 
 if have_srp
     platform_files += 'src/proto/eap.c'
-	platform_files += 'src/crypto/srp_constants.c'
-	platform_files += 'src/crypto/srp.c'
+    platform_files += 'src/crypto/srp_constants.c'
+    platform_files += 'src/crypto/srp.c'
 endif
 
+# Жестко отключаем HAVE_NETTLE, так как мы используем встроенные aes.c и sha256.c
+cdata.set10('HAVE_NETTLE', false)
 cdata.set10('HAVE_MBEDTLS', mbedcrypto_lib_found)
-cdata.set10('HAVE_NETTLE', use_gnutls and not mbedcrypto_lib_found)
 # Generate config.h
 config_file = configure_file(output: 'config.h', configuration: cdata)
 
-- 
2.52.0.windows.1

