From d068509aa0cc8731e6c9cc6c3ebb5152f9c428ac Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Fri, 5 Jun 2026 09:05:11 +0300
Subject: [PATCH] fix Shim-Loader header parser

---
 gen_shim.py | 27 +++++++++++++++++++++++----
 1 file changed, 23 insertions(+), 4 deletions(-)

diff --git a/gen_shim.py b/gen_shim.py
index f4416d1..b0fdd53 100755
--- a/gen_shim.py
+++ b/gen_shim.py
@@ -13,16 +13,36 @@ os.chdir(os.path.dirname(__file__))
 vkxml = "Vulkan-Headers/registry/vk.xml"
 
 commands = []
+# Явный белый список KHR функций, жизненно необходимых для lcevc
+LCEVC_REQUIRED_FUNCTIONS = {
+    "vkGetPhysicalDeviceSurfaceSupportKHR",
+    "vkDestroySwapchainKHR",
+    "vkDestroySurfaceKHR",
+    "vkGetPhysicalDeviceSurfaceCapabilitiesKHR",
+    "vkGetPhysicalDeviceSurfaceFormatsKHR",
+    "vkGetPhysicalDeviceSurfacePresentModesKHR",
+    "vkCreateSwapchainKHR",
+    "vkGetSwapchainImagesKHR"
+}
+
 for cmd in ET.parse(vkxml).getroot().findall(".//commands/command"):
     proto = cmd.find("proto")
     if proto is None:
         continue
-    if "vulkan" not in cmd.get("export", "").lower().split(","):
-        continue
 
     name = proto.find("name").text
-    # The XML is weird, stuff like "const" or "*" Are outside of the <type> tags.
+
+    if any(cmd_item[1] == name for cmd_item in commands):
+        continue
+
+    export_attr = cmd.get("export", "").lower().split(",")
+    is_vulkan_export = "vulkan" in export_attr
+
+    if not (is_vulkan_export or name in LCEVC_REQUIRED_FUNCTIONS):
+        continue
+
     ret_type = "".join(proto.itertext()).strip().rsplit(" ", 1)[0]
+    
     params = []
     param_names = []
     for p in cmd.findall("param"):
@@ -31,7 +51,6 @@ for cmd in ET.parse(vkxml).getroot().findall(".//commands/command"):
 
     commands.append((ret_type, name, params, param_names))
 
-
 print("""// Auto-Generated from vk.xml
 #define VK_NO_PROTOTYPES
 #include <vulkan/vulkan.h>
-- 
2.52.0.windows.1

