From f7770013e7084ba5c8a35a4bdc035acec8ca661b Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Sat, 25 Apr 2026 17:15:23 +0300
Subject: [PATCH] patch for 10.0

---
diff --git a/Source/API/SvtJpegxs.h b/Source/API/SvtJpegxs.h
index 4bdc178..e47fd39 100644
--- a/Source/API/SvtJpegxs.h
+++ b/Source/API/SvtJpegxs.h
@@ -20,6 +20,7 @@ extern "C" {
 #ifdef DEF_DLL
 #ifdef DEF_BUILDING_SHARED_LIBS
 #if defined(_WIN32)
+#undef PREFIX_API
 #define PREFIX_API __declspec(dllexport)
 #elif defined(__GNUC__) && __GNUC__ >= 4
 #define PREFIX_API __attribute__((visibility("default")))
@@ -28,6 +29,7 @@ extern "C" {
 #endif
 #else /*DEF_BUILDING_SHARED_LIBS*/
 #if defined(_WIN32)
+#undef PREFIX_API
 #define PREFIX_API __declspec(dllimport)
 #else /*other compiler*/
 #define PREFIX_API