From f7770013e7084ba5c8a35a4bdc035acec8ca661b Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Sat, 25 Apr 2026 17:15:23 +0300
Subject: [PATCH] patch for 10.0

---
diff --git a/Source/API/SvtJpegxs.h b/Source/API/SvtJpegxs.h
index 4bdc178..dc7d47c 100644
--- a/Source/API/SvtJpegxs.h
+++ b/Source/API/SvtJpegxs.h
@@ -34,7 +34,11 @@ extern "C" {
 #endif
 #endif /*DEF_BUILDING_SHARED_LIBS*/
 #else  /*DEF_DLL*/
+#if defined(__MINGW32__) || defined(_WIN32)
+#define PREFIX_API __declspec(dllexport)
+#else
 #define PREFIX_API
+#endif
 #endif /*DEF_DLL*/
 
 /********************************
 