From 576c80a0c03cac675fb4c9adf19c8d80a1769a28 Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Thu, 18 Jun 2026 09:53:01 +0300
Subject: [PATCH] clean repo

---
 CMakeLists.txt | 114 -------------------------------------------------
 1 file changed, 114 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 17d1dffa..c8e87f9b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -223,54 +223,6 @@ set(LIBDATACHANNEL_IMPL_HEADERS
 	${CMAKE_CURRENT_SOURCE_DIR}/src/impl/wshandshake.hpp
 )
 
-set(TESTS_SOURCES
-    ${CMAKE_CURRENT_SOURCE_DIR}/test/main.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/test/connectivity.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/test/negotiated.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/test/reliability.cpp
-	${CMAKE_CURRENT_SOURCE_DIR}/test/simulcast_sdp_generation.cpp
-	${CMAKE_CURRENT_SOURCE_DIR}/test/simulcast_sdp_parsing.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/test/turn_connectivity.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/test/track.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/test/video_layers_allocation.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/test/capi_connectivity.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/test/capi_track.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/test/websocket.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/test/websocketserver.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/test/capi_websocketserver.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/test/benchmark.cpp
-)
-
-if(NOT NO_MEDIA)
-    list(APPEND TESTS_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/test/fir.cpp)
-    list(APPEND TESTS_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/test/rtx.cpp)
-    list(APPEND TESTS_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/test/rtcp_app.cpp)
-endif()
-
-set(TESTS_HEADERS 
-	${CMAKE_CURRENT_SOURCE_DIR}/test/test.hpp
-)
-
-set(TESTS_UWP_RESOURCES
-	${CMAKE_CURRENT_SOURCE_DIR}/test/uwp/tests/Logo.png
-	${CMAKE_CURRENT_SOURCE_DIR}/test/uwp/tests/package.appxManifest
-	${CMAKE_CURRENT_SOURCE_DIR}/test/uwp/tests/SmallLogo.png
-	${CMAKE_CURRENT_SOURCE_DIR}/test/uwp/tests/SmallLogo44x44.png
-	${CMAKE_CURRENT_SOURCE_DIR}/test/uwp/tests/SplashScreen.png
-	${CMAKE_CURRENT_SOURCE_DIR}/test/uwp/tests/StoreLogo.png
-	${CMAKE_CURRENT_SOURCE_DIR}/test/uwp/tests/Windows_TemporaryKey.pfx
-)
-
-set(BENCHMARK_UWP_RESOURCES
-	${CMAKE_CURRENT_SOURCE_DIR}/test/uwp/benchmark/Logo.png
-	${CMAKE_CURRENT_SOURCE_DIR}/test/uwp/benchmark/package.appxManifest
-	${CMAKE_CURRENT_SOURCE_DIR}/test/uwp/benchmark/SmallLogo.png
-	${CMAKE_CURRENT_SOURCE_DIR}/test/uwp/benchmark/SmallLogo44x44.png
-	${CMAKE_CURRENT_SOURCE_DIR}/test/uwp/benchmark/SplashScreen.png
-	${CMAKE_CURRENT_SOURCE_DIR}/test/uwp/benchmark/StoreLogo.png
-	${CMAKE_CURRENT_SOURCE_DIR}/test/uwp/benchmark/Windows_TemporaryKey.pfx
-)
-
 if(RTC_UPDATE_VERSION_HEADER)
 	configure_file (
 		${PROJECT_SOURCE_DIR}/cmake/version.h.in
@@ -572,69 +524,3 @@ install(FILES
 	${CMAKE_BINARY_DIR}/LibDataChannelConfigVersion.cmake
     DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/LibDataChannel)
 
-# Tests
-if(NOT NO_TESTS)
-	if(CMAKE_SYSTEM_NAME STREQUAL "WindowsStore")
-		# Add resource files needed for UWP apps.
-		add_executable(datachannel-tests ${TESTS_SOURCES} ${TESTS_HEADERS} ${TESTS_UWP_RESOURCES})
-	else()
-		add_executable(datachannel-tests ${TESTS_SOURCES} ${TESTS_HEADERS})
-	endif()
-
-	set_target_properties(datachannel-tests PROPERTIES
-		VERSION ${PROJECT_VERSION}
-		CXX_STANDARD 17
-		OUTPUT_NAME tests)
-
-	set_target_properties(datachannel-tests PROPERTIES
-		XCODE_ATTRIBUTE_PRODUCT_BUNDLE_IDENTIFIER com.github.paullouisageneau.libdatachannel.tests)
-
-	target_include_directories(datachannel-tests PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/src)
-	target_link_libraries(datachannel-tests datachannel Threads::Threads)
-
-	# Benchmark
-	if(CMAKE_SYSTEM_NAME STREQUAL "WindowsStore")
-		# Add resource files needed for UWP apps.
-		add_executable(datachannel-benchmark test/benchmark.cpp ${BENCHMARK_UWP_RESOURCES})
-	else()
-		add_executable(datachannel-benchmark test/benchmark.cpp)
-	endif()
-
-	set_target_properties(datachannel-benchmark PROPERTIES
-		VERSION ${PROJECT_VERSION}
-		CXX_STANDARD 17
-		OUTPUT_NAME benchmark)
-
-	set_target_properties(datachannel-benchmark PROPERTIES
-		XCODE_ATTRIBUTE_PRODUCT_BUNDLE_IDENTIFIER com.github.paullouisageneau.libdatachannel.benchmark)
-
-	target_compile_definitions(datachannel-benchmark PRIVATE BENCHMARK_MAIN=1)
-	target_include_directories(datachannel-benchmark PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/src)
-	target_link_libraries(datachannel-benchmark datachannel Threads::Threads)
-endif()
-
-# Examples
-if(NOT NO_EXAMPLES)
-	set(JSON_BuildTests OFF CACHE INTERNAL "")
-	if(USE_SYSTEM_JSON)
-		find_package(nlohmann_json REQUIRED)
-	else()
-		add_subdirectory(deps/json EXCLUDE_FROM_ALL)
-	endif()
-
-	if(NOT NO_WEBSOCKET)
-		add_subdirectory(examples/client)
-		add_subdirectory(examples/client-benchmark)
-	endif()
-	if(NOT NO_MEDIA)
-		add_subdirectory(examples/media-receiver)
-		add_subdirectory(examples/media-sender)
-		add_subdirectory(examples/media-sfu)
-	endif()
-	if(NOT NO_MEDIA AND NOT NO_WEBSOCKET)
-    	add_subdirectory(examples/streamer)
-	endif()
-	add_subdirectory(examples/copy-paste)
-	add_subdirectory(examples/copy-paste-capi)
-endif()
-
-- 
2.52.0.windows.1

