From d4ded0d0154e233bec4dc6ab63f6105e8182a347 Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Wed, 17 Jun 2026 10:40:38 +0300
Subject: [PATCH] disable git version detect

---
 CMakeLists.txt          | 13 -------------
 avs_core/CMakeLists.txt | 14 --------------
 2 files changed, 27 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 9ec349bd..437238cb 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -568,19 +568,6 @@ else()
   target_architecture(AVS_ARCH)
   CONFIGURE_FILE(${CMAKE_CURRENT_SOURCE_DIR}/avs_core/core/arch.h.in ${CMAKE_CURRENT_BINARY_DIR}/arch.h @ONLY)
 
-  # Dynamically generate the sequential version info from Git
-  # Based on the example here: http://www.cmake.org/pipermail/cmake/2010-July/038015.html
-  FIND_PACKAGE(Git)
-  INCLUDE_DIRECTORIES(${CMAKE_CURRENT_BINARY_DIR})
-  ADD_CUSTOM_TARGET(
-    VersionGen ALL
-    ${CMAKE_COMMAND} -D SRC=${CMAKE_CURRENT_SOURCE_DIR}/avs_core/core/version.h.in
-                      -D DST=${CMAKE_CURRENT_BINARY_DIR}/version.h
-                      -D GIT=${GIT_EXECUTABLE}
-                      -D REPO=${CMAKE_SOURCE_DIR}
-                      -P ${CMAKE_CURRENT_SOURCE_DIR}/avs_core/Version.cmake
-  )
-
   include(GNUInstallDirs)
 
   install(
diff --git a/avs_core/CMakeLists.txt b/avs_core/CMakeLists.txt
index 346c9350..529be008 100644
--- a/avs_core/CMakeLists.txt
+++ b/avs_core/CMakeLists.txt
@@ -318,20 +318,6 @@ include("${CMAKE_CURRENT_LIST_DIR}/TargetArch.cmake")
 target_architecture(AVS_ARCH)
 CONFIGURE_FILE(${CMAKE_CURRENT_SOURCE_DIR}/core/arch.h.in ${CMAKE_CURRENT_BINARY_DIR}/arch.h @ONLY)
 
-# Dynamically generate the sequential version info from Git
-# Based on the example here: http://www.cmake.org/pipermail/cmake/2010-July/038015.html
-FIND_PACKAGE(Git)
-INCLUDE_DIRECTORIES(${CMAKE_CURRENT_BINARY_DIR})
-ADD_CUSTOM_TARGET(
-    VersionGen
-    ${CMAKE_COMMAND} -D SRC=${CMAKE_CURRENT_SOURCE_DIR}/core/version.h.in
-                     -D DST=${CMAKE_CURRENT_BINARY_DIR}/version.h
-                     -D GIT=${GIT_EXECUTABLE}
-                     -D REPO=${CMAKE_SOURCE_DIR}
-                     -P ${CMAKE_CURRENT_SOURCE_DIR}/Version.cmake
-)
-ADD_DEPENDENCIES("AvsCore" VersionGen)
-
 IF(MSVC_IDE)
   add_custom_command(
     TARGET AvsCore
-- 
2.52.0.windows.1

