From 33ba026f390bd6dff0db9d6cb07f9e3f3b868cb4 Mon Sep 17 00:00:00 2001
From: lifegpc <g1710431395@gmail.com>
Date: Wed, 30 Jun 2021 21:20:27 +0800
Subject: [PATCH] Fix missing LT_API which required by ffmpeg

---
 libtwolame/twolame.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/libtwolame/twolame.h b/libtwolame/twolame.h
index 6b3ee9d..06f86cc 100644
--- a/libtwolame/twolame.h
+++ b/libtwolame/twolame.h
@@ -263,7 +263,7 @@ TL_API int twolame_encode_buffer_float32(twolame_options * glopts,
  *  \return                The number of bytes put in output buffer
  *                         or a negative value on error
  */
-int twolame_encode_buffer_float32_interleaved(twolame_options * glopts,
+TL_API int twolame_encode_buffer_float32_interleaved(twolame_options * glopts,
         const float pcm[],
         int num_samples,
         unsigned char *mp2buffer, int mp2buffer_size);
