From 6fced852d4d5cfad58cf9dbe3ea619b08e87d398 Mon Sep 17 00:00:00 2001
From: Nicolas PARLANT <nicolas.parlant@parhuet.fr>
Date: Tue, 24 Feb 2026 12:03:41 +0100
Subject: [PATCH] fix old-style-def for c23 (#105)

Signed-off-by: Nicolas PARLANT <nicolas.parlant@parhuet.fr>
---
 libtwolame/psycho_2.c | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/libtwolame/psycho_2.c b/libtwolame/psycho_2.c
index 07363d1..806b9b7 100644
--- a/libtwolame/psycho_2.c
+++ b/libtwolame/psycho_2.c
@@ -54,9 +54,7 @@ static const FLOAT bmax[27] = { 20.0, 20.0, 20.0, 20.0, 20.0, 17.0, 15.0,
                                 4.5, 4.5, 4.5, 3.5, 3.5, 3.5
                               };
 
-static void psycho_2_read_absthr(absthr, table)
-FLOAT *absthr;
-int table;
+static void psycho_2_read_absthr(FLOAT *absthr, int table)
 {
     int j;
 #include "psycho_2_absthr.h"
