From 4f74abbdd8f4ff68943815488d8f891b7624c66d Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Sun, 26 Apr 2026 21:37:19 +0300
Subject: [PATCH] fix2

---
 src_main/CMakeLists.txt | 24 ++++++++++++++++++------
 1 file changed, 18 insertions(+), 6 deletions(-)

diff --git a/src_main/CMakeLists.txt b/src_main/CMakeLists.txt
index 1e4fddf..6adf3a4 100644
--- a/src_main/CMakeLists.txt
+++ b/src_main/CMakeLists.txt
@@ -56,13 +56,25 @@ endif()
 
 if(XEVD_INSTALL)
     include(GNUInstallDirs)
+
+    set_target_properties(xevd PROPERTIES PUBLIC_HEADER "../../inc/xevd.h")
+
     install(TARGETS xevd xevd_shared
             EXPORT XEVDTargets
-            ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
-            LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
-            RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
-            PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
-            
+            ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR} COMPONENT Development
+            LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR} COMPONENT Libraries
+            RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR} COMPONENT Libraries
+            PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR} COMPONENT Development
+    )
+
     configure_file("${CMAKE_CURRENT_SOURCE_DIR}/../pkgconfig/xevd.pc.in" "${CMAKE_CURRENT_BINARY_DIR}/xevd.pc" @ONLY)
-    install(FILES "${CMAKE_CURRENT_BINARY_DIR}/xevd.pc" DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig)
+    install(FILES "${CMAKE_CURRENT_BINARY_DIR}/xevd.pc" 
+            DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig 
+            COMPONENT Development)
+
+    install(EXPORT XEVDTargets
+            FILE XEVDTargets.cmake
+            NAMESPACE XEVD::
+            DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/XEVD
+            COMPONENT Development)
 endif()
-- 
2.52.0.windows.1

