From 0bbdfcf81db35fdcbadfedb07e3d7d067331be10 Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Fri, 19 Jun 2026 11:59:57 +0300
Subject: [PATCH] fix directx Descriptor Handle

---
 src/libutils/oglapphelpers/hlsl.cpp | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/src/libutils/oglapphelpers/hlsl.cpp b/src/libutils/oglapphelpers/hlsl.cpp
index e165e420..fd87245f 100644
--- a/src/libutils/oglapphelpers/hlsl.cpp
+++ b/src/libutils/oglapphelpers/hlsl.cpp
@@ -395,10 +395,14 @@ void HLSLBuilder::allocateAllTextures()
         srvDesc.ViewDimension = D3D12_SRV_DIMENSION_TEXTURE3D;
         srvDesc.Texture3D.MipLevels = 1;
 
+        D3D12_CPU_DESCRIPTOR_HANDLE baseHandle;
+        m_cbvSrvHeap->GetCPUDescriptorHandleForHeapStart(&baseHandle);
+
         CD3DX12_CPU_DESCRIPTOR_HANDLE srvHandle(
-            m_cbvSrvHeap->GetCPUDescriptorHandleForHeapStart(),
+            baseHandle,
             heapSlot,
             m_srvDescriptorSize);
+        // -----------------------------------------------------
 
         m_device->CreateShaderResourceView(texRes.m_texture.Get(), &srvDesc, srvHandle);
 
@@ -457,10 +461,14 @@ void HLSLBuilder::allocateAllTextures()
             srvDesc.Texture2D.MipLevels = 1;
         }
 
+        D3D12_CPU_DESCRIPTOR_HANDLE baseHandle2;
+        m_cbvSrvHeap->GetCPUDescriptorHandleForHeapStart(&baseHandle2);
+
         CD3DX12_CPU_DESCRIPTOR_HANDLE srvHandle(
-            m_cbvSrvHeap->GetCPUDescriptorHandleForHeapStart(),
+            baseHandle2,
             heapSlot,
             m_srvDescriptorSize);
+        // -----------------------------------------------------
 
         m_device->CreateShaderResourceView(texRes.m_texture.Get(), &srvDesc, srvHandle);
 
-- 
2.52.0.windows.1

