From 9ae13c9b829cd4fc544b8ee0d138d9e78d171549 Mon Sep 17 00:00:00 2001
From: Alexpux <ccom@randomderp.com>
Date: Mon, 14 Sep 2020 02:40:17 +0000
Subject: [PATCH] dont-use-outdated-symbol-list

---
diff --git i/libmp3lame/Makefile.am w/libmp3lame/Makefile.am
index 3581f4ed..a9bca385 100644
--- i/libmp3lame/Makefile.am
+++ w/libmp3lame/Makefile.am
@@ -22,7 +22,6 @@ endif
 libmp3lame_la_LIBADD =	$(cpu_ldadd) $(vector_ldadd) $(decoder_ldadd) \
 			@LIBMP3LAME_LDADD@ $(CONFIG_MATH_LIB)
 libmp3lame_la_LDFLAGS = -version-info @LIB_MAJOR_VERSION@:@LIB_MINOR_VERSION@ \
-			-export-symbols $(top_srcdir)/include/libmp3lame.sym \
 			-no-undefined
 
 INCLUDES = @INCLUDES@ -I$(top_srcdir) -I$(top_builddir)
