From 8e67cbed0d856a84e2d1c15a0ddf1539b8486629 Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Thu, 23 Apr 2026 05:02:49 +0300
Subject: [PATCH] Ignore not used params

---
--- a/Makefile.am
+++ b/Makefile.am
@@ -10,7 +10,7 @@
 ACLOCAL_AMFLAGS=-I m4
 
 .PHONY: test
 
-SUBDIRS = mpglib libmp3lame frontend Dll doc include misc dshow ACM \
+SUBDIRS = libmp3lame include
 	mac vc_solution
 
 CLEANFILES = testcase.new.mp3
 
 EXTRA_DIST = \
-- 
2.52.0.windows.1
