From 5796623cf42fc295adec35f0409e2d616a333e04 Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Tue, 16 Jun 2026 19:20:02 +0300
Subject: [PATCH] disable git version detect

---
 configure | 2 --
 1 file changed, 2 deletions(-)

diff --git a/configure b/configure
index 8e836c00..59ceb58d 100755
--- a/configure
+++ b/configure
@@ -1901,8 +1901,6 @@ if [ "$static" = "yes" ]; then
     echo 'install: install-lib-static' >> config.mak
 fi
 
-${SRCPATH}/version.sh >> x264_config.h
-
 if [ "$bashcompletion" = "auto" ]; then
     if [ "$cli" = "no" ]; then
         bashcompletion="no"
-- 
2.52.0.windows.1

