From f7770013e7084ba5c8a35a4bdc035acec8ca661b Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Sat, 25 Apr 2026 17:15:23 +0300
Subject: [PATCH] Accept SVT-HEVC 1.5.1

---
 source/cmake/Findsvthevc.cmake | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/source/cmake/Findsvthevc.cmake b/source/cmake/Findsvthevc.cmake
index 712beba6d..137964c91 100644
--- a/source/cmake/Findsvthevc.cmake
+++ b/source/cmake/Findsvthevc.cmake
@@ -16,7 +16,7 @@ SET(CMAKE_FIND_LIBRARY_SUFFIXES ".lib")
 endif()
 
 set(SVT_VERSION_MAJOR_REQUIRED       1)
-set(SVT_VERSION_MINOR_REQUIRED       4)
+set(SVT_VERSION_MINOR_REQUIRED       5)
 set(SVT_VERSION_PATCHLEVEL_REQUIRED  1)
 
 find_path(SVT_HEVC_INCLUDE_DIR
@@ -38,7 +38,7 @@ if(SVT_HEVC_INCLUDE_DIR)
         string(REGEX MATCH "SVT_VERSION_PATCHLEVEL  \\(([0-9]*)\\)" _ ${version})
         set(SVT_VERSION_PATCHLEVEL ${CMAKE_MATCH_1})
 
-        if(NOT ${SVT_VERSION_MAJOR} EQUAL "1" OR NOT ${SVT_VERSION_MINOR} EQUAL "4" OR NOT ${SVT_VERSION_PATCHLEVEL} EQUAL "1")
+        if(NOT ${SVT_VERSION_MAJOR} EQUAL "1" OR NOT ${SVT_VERSION_MINOR} EQUAL "5" OR NOT ${SVT_VERSION_PATCHLEVEL} EQUAL "1")
             message (SEND_ERROR "-- Found SVT-HEVC Lib Version: ${SVT_VERSION_MAJOR}.${SVT_VERSION_MINOR}.${SVT_VERSION_PATCHLEVEL} which doesn't match the required version: ${SVT_VERSION_MAJOR_REQUIRED}.${SVT_VERSION_MINOR_REQUIRED}.${SVT_VERSION_PATCHLEVEL_REQUIRED}; Aborting configure  ")
         else()
             message(STATUS "-- Found SVT-HEVC Lib Version: ${SVT_VERSION_MAJOR}.${SVT_VERSION_MINOR}.${SVT_VERSION_PATCHLEVEL}")
-- 
2.52.0.windows.1

