From 5e2042e86a4946e9eb9241cca13d79045b79162a Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Sat, 9 May 2026 10:02:20 +0300
Subject: [PATCH] add more langs for flite

---
 libavfilter/asrc_flite.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/libavfilter/asrc_flite.c b/libavfilter/asrc_flite.c
index a7471607cf..5f12bd32a2 100644
--- a/libavfilter/asrc_flite.c
+++ b/libavfilter/asrc_flite.c
@@ -46,6 +46,14 @@ DECLARE_REGISTER_VOICE_FN(kal16);
 DECLARE_REGISTER_VOICE_FN(rms);
 DECLARE_REGISTER_VOICE_FN(slt);
 
+/* Declaring functions for voices with non-standard prefixes */
+cst_voice *register_cmu_time_awb(const char *);
+void     unregister_cmu_time_awb(cst_voice *);
+cst_voice *register_cmu_indic_lang(const char *);
+void     unregister_cmu_indic_lang(cst_voice *);
+cst_voice *register_cmu_grapheme_lang(const char *);
+void     unregister_cmu_grapheme_lang(cst_voice *);
+
 struct voice_entry {
     const char *name;
     cst_voice * (*register_fn)(const char *);
@@ -66,6 +74,23 @@ static struct voice_entry voice_entries[] = {
     MAKE_VOICE_STRUCTURE(kal16),
     MAKE_VOICE_STRUCTURE(rms),
     MAKE_VOICE_STRUCTURE(slt),
+    
+ /* Add new voices manually, bypassing the hard macro */
+    {
+        .name          = "time_awb",
+        .register_fn   = register_cmu_time_awb,
+        .unregister_fn = unregister_cmu_time_awb,
+    },
+    {
+        .name          = "indic_lang",
+        .register_fn   = register_cmu_indic_lang,
+        .unregister_fn = unregister_cmu_indic_lang,
+    },
+    {
+        .name          = "grapheme_lang",
+        .register_fn   = register_cmu_grapheme_lang,
+        .unregister_fn = unregister_cmu_grapheme_lang,
+    },
 };
 
 static void list_voices(void *log_ctx, const char *sep)
-- 
2.52.0.windows.1

