From 7edbc2131b59fd171d7142b6978bccd56867e8a7 Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Fri, 29 May 2026 21:35:55 +0300
Subject: [PATCH] fix return type defaults to 'int' for ia_mpegh

---
 libavcodec/ia_mpegh_enc.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/libavcodec/ia_mpegh_enc.c b/libavcodec/ia_mpegh_enc.c
index 12241fe313..55bfa4144f 100644
--- a/libavcodec/ia_mpegh_enc.c
+++ b/libavcodec/ia_mpegh_enc.c
@@ -1035,7 +1035,7 @@ void ia_mpegh_handle_hoa(ia_input_config *pstr_input_config,IA_MPEGH_Context *ct
 	pstr_input_config->aud_ch_pcm_cfg.sample_rate = g_inp_hoa_config[0].sample_rate;
 }
 //
-static av_cold ia_mpegh_encode_init(AVCodecContext *avctx)
+static av_cold int ia_mpegh_encode_init(AVCodecContext *avctx)
 {
 	IA_MPEGH_Context *ctx = avctx->priv_data;
 
@@ -1117,7 +1117,7 @@ static av_cold ia_mpegh_encode_init(AVCodecContext *avctx)
 	return err_code;
 }
 
-static av_cold ia_mpegh_encode_close(AVCodecContext *avctx)
+static av_cold int ia_mpegh_encode_close(AVCodecContext *avctx)
 {
 	IA_MPEGH_Context *ctx = avctx->priv_data;
 
-- 
2.52.0.windows.1

