From 9eff9cb08f4846fcebd7b229d038471863e274d9 Mon Sep 17 00:00:00 2001
From: netony <hevc1@foxmail.com>
Date: Sun, 7 Jun 2026 00:33:40 +0800
Subject: [PATCH] Update libx264.c

---
 libavcodec/libx264.c | 28 ++++++++++++++--------------
 1 file changed, 14 insertions(+), 14 deletions(-)

diff --git a/libavcodec/libx264.c b/libavcodec/libx264.c
index b688d4f321841..c4757058cb274 100644
--- a/libavcodec/libx264.c
+++ b/libavcodec/libx264.c
@@ -1156,10 +1156,10 @@ static av_cold int X264_init(AVCodecContext *avctx)
     if (x4->scenechange_threshold >= 0)
         x4->params.i_scenecut_threshold = x4->scenechange_threshold;
 
-    if (avctx->qmin >= 0)
-        x4->params.rc.i_qp_min          = avctx->qmin;
-    if (avctx->qmax >= 0)
-        x4->params.rc.i_qp_max          = avctx->qmax;
+    // if (avctx->qmin >= 0)
+    //     x4->params.rc.i_qp_min          = avctx->qmin;
+    // if (avctx->qmax >= 0)
+    //     x4->params.rc.i_qp_max          = avctx->qmax;
     if (avctx->max_qdiff >= 0)
         x4->params.rc.i_qp_step         = avctx->max_qdiff;
     if (avctx->qblur >= 0)
@@ -1292,16 +1292,16 @@ static av_cold int X264_init(AVCodecContext *avctx)
     if (x4->b_frame_strategy >= 0)
         x4->params.i_bframe_adaptive = x4->b_frame_strategy;
 
-    if (x4->profile)
-        if (x264_param_apply_profile(&x4->params, x4->profile) < 0) {
-            int i;
-            av_log(avctx, AV_LOG_ERROR, "Error setting profile %s.\n", x4->profile);
-            av_log(avctx, AV_LOG_INFO, "Possible profiles:");
-            for (i = 0; x264_profile_names[i]; i++)
-                av_log(avctx, AV_LOG_INFO, " %s", x264_profile_names[i]);
-            av_log(avctx, AV_LOG_INFO, "\n");
-            return AVERROR(EINVAL);
-        }
+    // if (x4->profile)
+    //     if (x264_param_apply_profile(&x4->params, x4->profile) < 0) {
+    //         int i;
+    //         av_log(avctx, AV_LOG_ERROR, "Error setting profile %s.\n", x4->profile);
+    //         av_log(avctx, AV_LOG_INFO, "Possible profiles:");
+    //         for (i = 0; x264_profile_names[i]; i++)
+    //             av_log(avctx, AV_LOG_INFO, " %s", x264_profile_names[i]);
+    //         av_log(avctx, AV_LOG_INFO, "\n");
+    //         return AVERROR(EINVAL);
+    //     }
 
     x4->params.i_width          = avctx->width;
     x4->params.i_height         = avctx->height;
