From 9a5ea3ea90c6b9168df77d8fb0efa3c18ec84a7a Mon Sep 17 00:00:00 2001
From: Lichtenshtein <altruumann@protonmail.com>
Date: Tue, 21 Jul 2026 21:43:01 +0300
Subject: [PATCH] load avisynth plugins from subdir

---
 libavformat/avisynth.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/libavformat/avisynth.c b/libavformat/avisynth.c
index cb2be10..349ee49 100644
--- a/libavformat/avisynth.c
+++ b/libavformat/avisynth.c
@@ -180,6 +180,14 @@ static av_cold int avisynth_context_create(AVFormatContext *s)
             return ret;
 
     avs->env = avs->avs_library.avs_create_script_environment(3);
+#ifdef _WIN32
+    if (avs->env) {
+        AVS_Value path_val = avs_new_value_string("PROGRAMDIR\\avisynth");
+        AVS_Value args[1] = { path_val }; 
+        AVS_Value res = avs->avs_library.avs_invoke(avs->env, "AddAutoloadDir", avs_new_value_array(args, 1), NULL);
+        avs->avs_library.avs_release_value(res);
+    }
+#endif
     if (avs->avs_library.avs_get_error) {
         const char *error = avs->avs_library.avs_get_error(avs->env);
         if (error) {
-- 
2.52.0.windows.1

